OpenRA.Mods.Common/Traits/Buildings/RemoteDockHost.cs

```diff
@@ -0,0 +1,226 @@
+		}
+
+		public virtual bool Reserve(Actor self, DockClientManager client)
+		{
+			if (CanBeReserved && !ReservedDockClients.Contains(client))
+			{
+				ReservedDockClients.Add(client);
+				client.ReserveHost(self, this);
+				return true;
+			}
+
+			return false;
+		}
+
+		public virtual void UnreserveAll()
+		{
+			while (ReservedDockClients.Count > 0)
+				Unreserve(ReservedDockClients[0]);
+		}
+
+		public virtual void Unreserve(DockClientManager client)
+		{
+			if (ReservedDockClients.Contains(client))
+			{
+				ReservedDockClients.Remove(client);
+				client.UnreserveHost();
+			}
+		}
+
+		public virtual void OnDockStarted(Actor self, Actor clientActor, DockClientManager client)
+		{
+			if (Info.OccupyClient || DockWait <= 0)
+				DockStarted(self, new TraitPair<DockClientManager>(clientActor, client));
+			else
+				WaitingClients.Add((new TraitPair<DockClientManager>(clientActor, client), DockWait + self.World.WorldTick));
+		}
+
+		public virtual void OnDockCompleted(Actor self, Actor clientActor, DockClientManager client)
+		{
+			if (clientActor != null && !clientActor.IsDead && clientActor.IsInWorld)
+				client.OnDockCompleted(clientActor, self, this);
+
+			DockedClients.Remove(DockedClients.First(c => c.Trait == client));
+		}
+
+		public virtual bool QueueMoveActivity(Activity moveToDockActivity, Actor self, Actor clientActor, DockClientManager client)
+		{
+			if ((clientActor.CenterPosition - DockPosition).HorizontalLengthSquared > Info.Range.LengthSquared)
+			{
+				// TODO: MoveWithinRange doesn't support offsets.
+				moveToDockActivity.QueueChild(clientActor.Trait<IMove>().MoveWithinRange(Target.FromActor(self), Info.Range));
+				return true;
+			}
+
+			return false;
+		}
+
+		public virtual void QueueDockActivity(Activity moveToDockActivity, Actor self, Actor clientActor, DockClientManager client)
+		{
+			if (Info.OccupyClient)
+			{
+				if (moveToDockActivity == null)
+					clientActor.QueueActivity(new RemoteDockSequence(client, self, this, Info.Range.Length));
+				else
+					moveToDockActivity.QueueChild(new RemoteDockSequence(client, self, this, Info.Range.Length));
+			}
+			else
+				OnDockStarted(self, clientActor, client);
+		}
+
+		#endregion
+
+		void ITick.Tick(Actor self)
+		{
+			Tick(self);
+		}
+
+		protected virtual void Tick(Actor self)
+		{
+			// We are not allowed to use activities.
+			if (!Info.OccupyClient)
+			{
+				// Track wait time manually.
+				for (var i = 0; i < WaitingClients.Count; i++)
+				{
+					var clientActor = WaitingClients[i].Client.Actor;
+					var client = WaitingClients[i].Client.Trait;
+
+					if (WaitingClients[i].Time < self.World.WorldTick)
+					{
+						if (clientActor != null && !clientActor.IsDead && clientActor.IsInWorld && !IsTraitDisabled && !preventDock
+							&& (clientActor.CenterPosition - DockPosition).HorizontalLengthSquared <= Info.Range.LengthSquared
```

Comment:
This would also benefit from a helper method. Move everything except the `preventDock` check into it and then re-use it below as well.